src/main/scala/lms/core/codegen/C_codegen.scala

```diff
@@ -148,7 +148,7 @@ class ExtendedCCodeGen extends CompactCodeGen with ExtendedCodeGen {
     }
     withWraper(wraper _)(f)
   }
-  def quoteBlock(b: Block): Unit = quoteTypedBlock(b, true, false)
+  def quoteBlock(b: Block): Unit = ???
```

Comment:
in c code gen, it should not be allowed to generate closure. 
this is moved to cpp code gen.

Reply:
It depends ... gcc has support for nested functions. There should probably be some kind of flag to enable or disable gcc/clang specific patterns. We already use some other patterns like statement expressions that aren't part of ANSI C. It's nice to have them available because the generated code is more compact (and often more readable) but we have to be able to turn these off for less-fancy compilers (importantly, TCC).

In plain C we can add a closure conversion phase (I believe @GSAir implemented parts of it for OSR)

Not saying all this needs to be done right now, perhaps let's create a ticket to remind us later.

Reply:
Interesting! I didn't know that gcc support nested functions :) 

But so far, `quoteTypedBlock` goes to closure generation (via syntax: `[&](int a){ ... }), which needs to be fixed I think (we should use normal function syntax such as `int f(int x) {..}`

I also tested and found that the nested function can access variables defined out of the scope, such as ```
int main()
{
    int x = 5;
    int inner(int a) { 
        return a + x; 
    }
    printf("%d\n", inner(3));
    return 0;
}```
So this is not so different from closures IMO. Are there any difference between gcc-supported nested functions and c++ closures? 

It seems that there are a few things to handle here:
1. flag to support inner functions or not (decided based on if we use gcc or tcc)
2. when supporting inner functions, generate them in function syntax `int f(int x) { ... }`
3. when not supporting inner functions, closure conversions
4. for cpp codegen compiled with gcc, it seems that we can either use cpp closure or inner functions (not sure if there are differences between them). How do we choose?
@TiarkRompf @GSAir @Kraks

Reply:
When targeting modern C++, I think there is no need to generate nested functions (in fact it is not supported by GNU C++ according to its document).